### ISSUE TITLE
AArch64: binary encoding

### ISSUE THREAD

**Issue by jeanlego** [eclipse/omr#2588]
I was going through the ARM documentation and i've noticed that some encoding have some weirdness that I don't know how to encode in hex.

The table is stored @ https://github.com/CAS-Atlantic/AArch64-Encoding.git

the weirdness is the != 11111
[exerpt from https://github.com/CAS-Atlantic/AArch64-Encoding/blob/master/AArch64_ops.pdf]
![screenshot from 2018-05-25 14-22-37](https://user-images.githubusercontent.com/9909445/40557609-1ba5acd0-6027-11e8-9f9b-fbb5228bfbd1.png)

the operation will do imediate offset when == 11111 else register offset

how should we go about this

@0xdaryl  @JamesKingdon  @aarongraham9

**Comment by aarongraham9** [comment]
Adding @knn-k.

**Comment by JamesKingdon** [comment]
Doesn't it just take the register number to use for Xm (the post-index register) ? I suspect I'm misunderstanding the question.

**Comment by jeanlego** [comment]
yep those would be easy since our registers are from 0 to 30 (since R31 is restricted ) we can simply store the one with immediate offset as a separate op code.
I am under the assumption that whatever has to be defined is OR'rd with the binary encoding (hex).
Is this correct? if so how do we go about the != 0000. 
![screenshot from 2018-05-25 14-56-18](https://user-images.githubusercontent.com/9909445/40559044-cce60fe0-602b-11e8-8c92-079dd7f11093.png)

I don't know that we can use these operation at the moment but I was looking to add them for completeness and if changes are made we can use them, or we can leverage them in some way.

**Comment by JamesKingdon** [comment]
Oh right. So I assume that like the Rn and Rd bits, the immh bits will be 0s in the initial value and we'll OR in the necessary (non zero) value at encoding time. Probably add an assert to check that the value for immh really is !=0

**Comment by jeanlego** [comment]
sounds good to me! there are ~1000 ops in there and I was thinking we should keep all of them for completeness and it can be reused if we keep it complete and well commented.
I'm going through them to add relevant comments since the table did not say a lot about them.
this way we can only refer back to one table when needs be. I'm genarating OpBinary.cpp, AARCH64ops.hpp, AARCH64ops.ops and hopefully OProperties.cpp from this table

**Comment by JamesKingdon** [comment]
Sounds fantastic!
